### PR TITLE
[NNC] Add missing data type support for abs and frac

### PR DESCRIPTION
**Pull request by penguinwu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48679 [NNC] Add missing data type support for abs and frac**

Summary: This addresses the remaining problem reported in issue #48053

Data type supports for aten kernels in SimpleIREvaluator are not
consistent w/ aten::native library implementation. In SimpleIREvaluator,
  - only float/double are supported on aten::abs (integral types and half
are missing)
  - only float/double are supported on aten::frac (half are missing)

It is also not clear from kernel.cpp source code what are the expected
input data types for an aten kernel, leading to potential missing data
type issues down the road.

This commit addresses both issues in a limited way by
 - Added type promotion ops from half/integral input types to float
 - Added a skeleton support for some type checking for aten kernels,
   currently, only check for valid data types for frac and abs to limit
   the scope of the change; but the utility function can be used for
   consistently adding type checking for all aten functions

Known limitations:
 - abs support for integral types can be made more effective by invoking
 std::abs for integral tensors (currently kFabs maps to std::fabs).
 Since that change is a bit more involved (e.g., changing IntrinsicsOp
 kFabs to kAbs and other code generators accordingly), will leave it to
 another issue
 - other aten kernels may need similar type checking and some scrutiny
 on the use of promoteToFloat to detect invalid data types early on.
 That is also left for another issue

Test Plan: test_jit_fuser_te.test_unary_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25344839](https://our.internmc.facebook.com/intern/diff/D25344839)

### CODE REVIEW

**Comment by penguinwu** [comment]
@bertmaher This PR is really two parts. One is to address the specific problem in #48053 (which certain data types were missing for `abs` and `frac`; the other is to add a prototype for type checking on input operands of kernel functions. 

The second part is not strictly needed for #48053. I added it because when diagnosing the issue, I found it hard to reason about which input types should be supported (some may require type promotion to the type expected by `IntrinsicsOp`) and which types are not. Since such information was not explicit in `kernel.cpp`, it is hard to reason about which data types are still missing for a kernel function by reading the code (as opposed to by running tests).

I am not completely sure if the second part is worth the effort yet (as it could be a beginner's eye perspective). Would love to hear your opinion on this, and I have no problem removing the 2nd part if you recommend otherwise.

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/kernel.cpp:448]: Not for this diff but this reminds me that I'd love for tensorexpr::Dtype::scalar_type to just use c10::ScalarType instead of defining its own thing that's kind of an extension of it.  Maybe file an issue/task for this?

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/kernel.cpp:513]: IMO we should stop throwing exceptions for unsupported cases since the fallback is disabled; we either need to handle things properly at this point, or clean things out during tensorexpr_fuser.

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1249]: Is promotion to float necessary for cuda?  I guess it probably doesn't hurt anything...

File: torch/csrc/jit/tensorexpr/kernel.h
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/kernel.h:39]: The type constraint idea is pretty interesting!  I kind of want to sleep on it but it seems reasonable to me.  Do we have anything like it elsewhere in the system (@ZolotukhinM, do you know?)

**Review by bertmaher — COMMENTED** [review]
Looking good, the only real concern I have is the introduction of a new `throw unsupported_dtype()`.  I want to think a bit more about the type constraints but it seems like a good idea to me so far.

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/kernel.cpp]: Oh, a minor bit of tidiness here, could you format the #includes as above, like using <...> and including the full torch/csrc/... path to the jit headers?  I kind of suspect those will make the internal buck build happy too.

File: torch/csrc/jit/tensorexpr/kernel.h
Comment (asuhan) [reply torch/csrc/jit/tensorexpr/kernel.h:39]: I'd still like to express the values of the enum with left bit shifts instead of literal values, would be more readable.

I'm not aware of using anything like it in the rest of the system, but I'd allow it on concision grounds, it's not really difficult to read it.

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (asuhan) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1249]: You mean necessary for the LLVM back-end? Because I think it'd work in CUDA through implicit promotion in the generated CUDA C++, but it wouldn't in LLVM IR which doesn't have implicit promotions.

**Review by asuhan — CHANGES_REQUESTED** [review]
Mostly looks good, I'm asking for a small change in the enum.

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (penguinwu) [reply torch/csrc/jit/tensorexpr/kernel.cpp:448]: Opened #48784 to track this request. I do not completely understand the scope of this change or the rationale of introducing `tensorexpr::ScalarType` in the first place, feel free to revise the description in #48784 (and add appropriate labels, etc)

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (penguinwu) [reply torch/csrc/jit/tensorexpr/kernel.cpp:513]: 4 places in the original `kernel.cpp` throws `unsupported_dtype()` (i.e., `promoteDType`, `demoteDType`, `constant`, `bindInput`. There are two reasons for unsupported dtypes:
- **User errors**, i.e., users specify invalid data types for a kernel function (e.g., `torch.frac()` does not take floating-point as input
- **Out-of-sync data type support** compared to eager mode, i.e., data type support in the JIT will lag behind that of eager runtime from time to time. I suspect this may be the reality, for instance, `kernel.cpp` does not seem to be no support for complex and qint types, and support for half floating-point may be lacking too (in `test_jit_fuser_te.py`, `test_unsupported_dtypes()` seems to give way to which types were considered as unsupported right now). This may be the direct consequence of how PyTorch rollout supports for new kernels or new types (eager mode first, then torch script, JIT, other devices, etc).  If we expect PyTorch kernels to continue evolving, there may always be a catch-up game.

If unsupported data types do occur in reality, I (as a developer) much prefer that the detection is explicit. As a user though, I'd prefer the error messages could differentiate between my (user) errors and the JIT's internal errors. Right now, `checkTypes` does not differentiate these two types of errors. For instance, for `frac` it will report unsupported data types for integer types (which will be user errors), but it will also report unsupported types for complex (actually I am not sure whether it is user errors or internal errors).

Comment (penguinwu) [reply torch/csrc/jit/tensorexpr/kernel.cpp:513]: Regarding un-even data type support, even the eager mode does not seem to support all data types in one batch. For instance, `torch.f16` is supported for `frac()` but not `sin()` (unless `sin()` is designed not to take `f16`?). If that is the case, then the data type support in PyTorch would be even more fluid.

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (penguinwu) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1249]: I will check in a fix so that integral types do not require promotion to float for `abs`. 

However, promotion from half to float is probably still needed. Since f16 is not a native type for normal CPU backend (or `std::xxx` functions), unless we develop math libraries for `f16`, promoting to f32 would be the easiest implementation. But for devices w/ native support for `f16` (e.g., gpu and most accelerators), this promotion may kill the performance benefit of `f16`. 

BTW, when we lower from `torch::jit::Node` to `IntrinsicsOp` in `TensorExprKernel::compile()`, we are not supposed to be platform-specific, so we have to normalize to the safest implementation (that works for all platforms). 

I feel the right approach is to make the code-gen of IntrinsicsOps to handle Half to float promotion. NOTE: as eager runtime supports more fp16 in kernel functions, this issue will pop up more. (If so, can we do this right approach in a separate issue to limit the scope of this one?)

File: torch/csrc/jit/tensorexpr/kernel.h
Comment (penguinwu) [reply torch/csrc/jit/tensorexpr/kernel.h:39]: @asuhan Good idea. I will use shift in the enum instead.

@bertmaher From my reading of the JIT code-base so far, I feel missing support for data types, mismatched data types, or type promotions can be a hidden killer for managing complexity or performance. My gut feeling is that the lack of vigorous specification (what data types are we supposed to support) and checking of types may hurt us in the longer run. 

On the performance side, integral data width matters a lot more on SIMD/GPU/accelerators than on CPUs (which promotes narrower ints to default ints for free), so implicit integer promotion is another thing we need to watch out for on the performance side.

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (penguinwu) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: Indeed CI checking has reported the problem, and it has been fixed by a subsequent commit. (BTW, these header files seem to be inserted by my VSCode, I still don't know how they got there :<).

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/kernel.cpp:448]: That's a great description, thanks.  I think the primary reason for te::ScalarType is historical; we started developing out-of-tree, with a group of folks pretty unfamiliar with PyTorch, and later patched in full support for at::ScalarType.  I'm not sure where all we use the "extended" types like kHandle, but I think it'd be cleaner to not pack them into a te-internal ScalarType.  It'd definitely be a somewhat involved change, though limited to tensorexpr/.

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/kernel.cpp:513]: Right so you're absolutely correct that there are currently places in `kernel.cpp` that throw `unsupported_dtype` but I think they should go away and we should either (1) actually support that dtype, or (2) filter those cases out in the `tensorexpr_fuser` pass.  As it stands right now, we'll attempt to fuse things that aren't supported, fail, and then the user is completely blocked, even though eager mode could handle the case.

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1249]: > can we do this right approach in a separate issue to limit the scope of this one

Yeah I think it's totally fine to do the "right approach" in future diffs.

**Review by bertmaher — APPROVED** [review]
Lgtm, feel free to land once you make the shifts change.

**Comment by penguinwu** [comment]
@asuhan Pushed the fix to use bit shift, please review again and approve if everything looks fine.

@bertmaher I opened a new issue (#48860) that tracks implementing a native support for integral types for `abs` (instead of promoting to float). I was hoping to amend this enhancement to this PR, but the fix was more involved than I thought (still debugging it), and I dont want to delay this PR further.

**Review by asuhan — APPROVED** [review]

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (eellison) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1249]: Last time handling `Half` in kernel.cpp was brought up the consensus was it was better to do it later in backend specific logic. 

Could IREval just call the Half-Rewriter before attempting to execute to handle Half ? Similar to how the other backends handle half:  https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/llvm_codegen.cpp#L317

Comment (penguinwu) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1250]: Opened issue #48864 to track this enhancement. Thanks @eellison for the pointer to half-rewriter, I think letting codegen handles more types is the right approach.

**Comment by penguinwu** [comment]
The PR caused a CI failure for `frac` on `half` along the `LLVM` code generator path, still debugging the problem.

```
JIT session error: Symbols not found: { __truncdfhf2 }
Program aborted due to an unhandled Error:
Failed to materialize symbols: { wrapper }Aborted (core dumped)
```

**Comment by penguinwu** [comment]
@bertmaher Please review the latest update. It fixes a problem exposed by a CI failure. 
- The fix promotes `half` to `float` instead of `DefaultType` (which could be `double` and which triggers a problem in `llvm_codegen`) when handling `frac` in `kernel.cpp`. 
- Also renamed `promoteIntToFloat()` to `promoteIntToDefaultType()` to avoid confusion

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (eellison) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1250]: This doesn't seem like the best solution for abs -
in eager we're we run abs(input) 
Here in NNC we're calling int(fabs(float(input)) which is more work & likely slower. What is motivating this change ?

**Review by eellison — COMMENTED** [review]
Hmm, could we think of possible alternative implementations ? 

I think Half support could be added to IREvaluator by using the HalfRewriter. 

To add support for integral intrinsics (abs) - i think we want to do something like https://github.com/pytorch/pytorch/pull/48973/files#diff-b800236a0a3b78031df12c064de7b33e4a72f5d076a7353f3d8f5a60160b0e15R901

I think there are a good number of places already where we don't model type promotion explicitly enough in operator lowerings (i'm going to issue abt it soon), but are saved by the fact that we call demoteOutput() at the end of every `compute{}Operand`. I think we should try to making extra casts than eager does in the future. Computing with a different dtype can yield numerical differences and performing two extra casts isn't free.

File: torch/csrc/jit/tensorexpr/kernel.cpp
Comment (penguinwu) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1250]: I am working on a fix to add integers to `abs`. Still debugging it (as `SimpleIREvaluator` right now assumes intrinsic are always floating point types, `abs` will be the first one that truly supports integer types), but will submit a PR soon.

Comment (penguinwu) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1250]: Regarding allowing `half` for `IntrinsicsOps`, we had a discussion on it in earlier reviews. We all agree w/ each other. But I propose to fix it in a separate issue and have opened #48864 to track it (I can work on it after my current set of issues land).

And regarding supporting integers for `kAbs`, it is tracked in #48860 (which I have started working on it last Friday, but got delayed because of debugging the unexpected CI failure in the PR :<).

Comment (eellison) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1250]: Cool! yea i think we're all agreed on best solution ! I guess I was just wondering whether it would be better to separately fix the integral evaluation and the half-rewriting than adding the extra casts/potentially changing numerics in the short term. It would be nice to fix the IREvaluator but when supporting it leaks into effecting codegen that seems not great. I'm fine doing whatever though.

 The link I posted adds an integral intrinsic - https://github.com/pytorch/pytorch/pull/48973/files#diff-b800236a0a3b78031df12c064de7b33e4a72f5d076a7353f3d8f5a60160b0e15R901. You might try something similar for `abs`. You'll soon find that no CI failure is all that unexpected 😛

**Pull request closed without merging**
